lsp-mode.el

```diff
@@ -161,29 +194,48 @@ Optional arguments:
-       (interactive)
-       (let ((client (make-lsp--client
-                       :language-id (if ,language-id-fn ,language-id-fn #'(lambda (_) ,language-id))
-                       :send-sync #'lsp--stdio-send-sync
-                       :send-async #'lsp--stdio-send-async
-                       :new-connection (lsp--make-tcp-connection
+    `(progn
+       (lsp-define-whitelist-enable ,name ,get-root)
+       (lsp-define-whitelist-disable ,name ,get-root)
+       (lsp-define-interactive-client ,name)
+
+       (defun ,enable ()
+         ,docstring
+         (interactive)
+         (let ((client (make-lsp--client
+                        :language-id (if ,language-id-fn ,language-id-fn #'(lambda (_) ,language-id))
+                        :send-sync #'lsp--stdio-send-sync
+                        :send-async #'lsp--stdio-send-async
+                        :new-connection (lsp--make-tcp-connection
                                          ,(symbol-name name)
                                          ,command
                                          ,command-fn
                                          ,host ,port)
-                       :get-root ,get-root
-                       :ignore-regexps ,ignore-regexps)))
-         (unless lsp-mode
-           ,(when initialize
-              `(funcall ,initialize client))
-           (let ((root (funcall (lsp--client-get-root client))))
-             (if (lsp--should-start-p root)
-               (progn
-                 (lsp-mode 1)
-                 (lsp--start client ,extra-init-params))
-               (message "Not initializing project %s" root))))))))
+                        :get-root ,get-root
+                        :ignore-regexps ,ignore-regexps)))
+           (lsp--start-client-maybe client :extra-init-params ,extra-init-params :initialize ,initialize))))))
+
+(cl-defun lsp--start-client-maybe (client
+                                 &key extra-init-params initialize)
+ "Start a client if lsp-mode is currently inactive, and the whitelist/blacklist allows it."
+            (unless lsp-mode
+              (when initialize
+                 (funcall initialize client))
+              (let ((root (funcall (lsp--client-get-root client))))
+                (if (lsp--should-start-p root)
+                    (progn
+                      (lsp-mode 1)
+                      (lsp--start client extra-init-params))
+                  (if (and ask lsp-enable-whitelist-ask)
+                      (let ((question (format-message "Add %s to lsp-project-whitelist? " root)))
+                        (if (y-or-n-p question) ;; cannot use when , side effects problem
+                            (progn (message "About to customize lsp-project-whitelist")
+                                   (customize-save-variable 'lsp-project-whitelist (add-to-list 'lsp-project-whitelist root)))
+                          (setq lsp--project-whitelist-temp (add-to-list 'lsp--project-whitelist-temp root)))
```

Comment:
I still don't understand the purpose of the `lsp--project-whitelist-temp` variable? The `lsp--project-whitelist` already stores all relevant information.

Reply:
If the user is asked to add the project to the whitelist, and says no, the whitelist is not updated.

If they now open another file, that shares the same project root, the fact that it is temporarily allowed to start is stored in `lsp--project-whitelist-temp`.  Otherwise it needs to be enabled manually for the second file, and may create an additional session.

Basically it allows the whitelist override to operate at the project root level, rather than at the individual buffer level.

Reply:
I understand, but that is *REALLY* hacky. You would be asked everytime if you want to add this to the whitelist, that is a real shitty experience. 

I would really like the approach that you just call `lsp-whitelist-project-add` and if the project is added, you could automatically start the lsp client.

Reply:
Ok, that could work.

I guess as a server-developer I am out of the normal use-case on this stuff.

As I want to have fine control of whether the server starts or not, based on the state of development of the server.

Reply:
But why wouldn't your use case work with the `lsp-whitelist-project-add` function? As I understand, you want control, when you want a project to be started with the language server?

By call the `lsp-whitelist-project-add` you would enable the language server and by removing, you would stop it. Sorry if I'm not understanding something right :/ 
You could also stop the lsp-client if you remove the project from a whitelist.

Reply:
I agree. I originally had the complex approach, then introduced the add/remove stuff, to do only that.

But making it control the server too is a good idea.

Thanks.